Plutarch/Rational.hs

```diff
@@ -21,121 +22,131 @@ import Plutarch (
   plam,
   plet,
   pmatch,
+  pto,
   (#),
   (#$),
   type (:-->),
  )
-import Plutarch.Bool (PEq (..), POrd (..), pif)
+import Plutarch.Bool (PEq, POrd, pif, (#<), (#<=), (#==))
 import Plutarch.Builtin (
   PAsData,
   PBuiltinList,
-  PData,
-  PIsData (..),
+  PIsData,
   pasInt,
   pasList,
+  pdata,
   pforgetData,
+  pfromData,
  )
-import Plutarch.Integer (PInteger, PIntegral (pdiv, pmod))
-import Plutarch.List (PListLike (pcons, phead, pnil, ptail), pmap)
-import Plutarch.Pair (PPair (..))
-import Plutarch.Show (PShow (pshow'), pshow)
-import Plutarch.Trace (ptraceError)
-import Plutarch.Unsafe (punsafeCoerce)
+import Plutarch.Integer (PInteger, pdiv, pmod)
+import Plutarch.Lift (pconstant)
+import Plutarch.List (pcons, phead, pnil, ptail)
+import Plutarch.NonZero (PNonZero, pnonZero')
+import Plutarch.Pair (PPair (PPair))
+import Plutarch.Show (PShow, pshow, pshow')
+import Plutarch.TermCont (tcont, unTermCont)
+import Plutarch.Unsafe (punsafeCoerce, punsafeFrom)
 
 data PRational s
-  = PRational (Term s PInteger) (Term s PInteger)
+  = PRational (Term s PInteger) (Term s PNonZero)
   deriving stock (GHC.Generic)
-  deriving anyclass (Generic, PlutusType)
+  deriving anyclass (Generic, PEq)
+
+instance PlutusType PRational where
+  type PInner PRational b = (PInteger :--> PNonZero :--> b) :--> b
+  pcon' (PRational x y') = plet y' $ \y -> plam $ \f -> f # x # y
```

Comment:
I had to make sure the denominator was evaluated prior to building the lambda. Otherwise, you could make a `n / 0` and the error would only happen upon actually matching on the rational.

It's inefficient sometimes - so there's also `pratLazy` below.

Reply:
For this PR, should I just remove the `PlutusType` instance and have 2 construction functions? (one without the invariant check, and one with it)